src/pal/src/misc/miscpalapi.cpp

```diff
@@ -226,86 +234,156 @@ PAL_GetPALDirectoryA(
+}
+
+#if defined(HAVE_SYS_GETRANDOM)
+static
+BOOL
+PAL_RandomFromGetRandom(
+        IN OUT LPVOID lpBuffer,
+        IN DWORD dwLength)
+{
+    ssize_t length = (ssize_t)dwLength;
+    static BOOL sMissingGetRandom;
+
+    if (sMissingGetRandom)
+    {
+        return FALSE;
+    }
+
+    do
+    {
+        ssize_t res = syscall(SYS_getrandom, lpBuffer, length,
+                              GRND_NONBLOCK);
+        if (res == -1)
         {
-            DWORD offset = 0;
-            do
+            if (errno == ENOSYS)
             {
-                ssize_t n = read(rand_des, (BYTE*)lpBuffer + offset , dwLength - offset);
-                if (n == -1)
-                {
-                    if (errno == EINTR)
-                    {
-                        continue;
-                    }
-                    ASSERT("read() failed, errno:%d (%s)\n", errno, strerror(errno));
-
-                    break;
-                }
-
-                offset += n;
+                sMissingGetRandom = TRUE;
+                return FALSE;
+            }
+            else if (errno == EINTR)
+            {
+                continue;
+            }
+            else
+            {
+                return FALSE;
             }
-            while (offset != dwLength);
-
-            _ASSERTE(offset == dwLength);
-
-            close(rand_des);
         }
+
+        length -= res;
+        lpBuffer = (char *)lpBuffer + res;
     }
+    while (length);
+
+    return TRUE;
+}
+#endif // HAVE_SYS_GETRANDOM
 
-    if (!sInitializedMRand)
+VOID
+PALAPI
+PAL_Random(
+        IN OUT LPVOID lpBuffer,
+        IN DWORD dwLength)
+{
+    static BOOL sInitializedMRand;
+    BOOL needSrand = TRUE;
+
+    PERF_ENTRY(PAL_Random);
+    ENTRY("PAL_Random(lpBuffer=%p, dwLength=%d)\n", lpBuffer, dwLength);
+
+#if defined(HAVE_SYS_GETRANDOM)
+    if (PAL_RandomFromGetRandom(lpBuffer, dwLength))
     {
-        srand48(time(NULL));
-        sInitializedMRand = TRUE;
+        needSrand = FALSE;
+    }
+    else
+#endif // HAVE_SYS_GETRANDOM
+    if (PAL_RandomFromUrandom(lpBuffer, dwLength))
+    {
+        needSrand = FALSE;
     }
 
-    // always xor srand48 over the whole buffer to get some randomness
-    // in case /dev/urandom is not really random
-
-    for (i = 0; i < dwLength; i++)
+    if (needSrand)
     {
-        if (i % sizeof(long) == 0) {
-            num = mrand48();
+        long num;
+
+        if (!sInitializedMRand)
+        {
+            // FIXME: There's not enough entropy in time(NULL) to
+            // initialize the PRNG.  Might be a good idea to leverage
+            // other sources of entropy at this point.
+            srand48(time(NULL));
+            sInitializedMRand = TRUE;
         }
 
-        *(((BYTE*)lpBuffer) + i) ^= num;
-        num >>= 8;
+        for (DWORD i = 0; i < dwLength; i++)
+        {
+            if (i % sizeof(long) == 0) {
+                num = mrand48();
+            }
+
+            *(((BYTE*)lpBuffer) + i) ^= num;
```

Comment:
While looking at the corefx counterpart of this code, I noticed that this is wrong: I'm trusting getrandom()/urandom here, and only using mrand48() if those didn't produce anything. So this XORs with potentially uninitialized values, which is UB territory. I'm fixing this to memcpy() in sizeof(long) chunks instead.